designs/odcr.md

```diff
@@ -0,0 +1,408 @@
+
+```yaml
+name: c5.large
+offerings:
+  - price: ....
+    available: ....
+    requirements:
+      - key: karpenter.sh/capacity-type
+        operator: In
+        values: ["reserved"]
+      - key: topology.kubernetes.io/zone
+        operator: In
+        values: ["us-west-2a"]
+      - key: topology.k8s.aws/zone-id
+        operator: In
+        values: ["usw2-az1"]
+  - price: ....
+    available: ....
+    requirements:
+      - key: karpenter.sh/capacity-type
+        operator: In
+        values: ["on-demand"]
+      - key: topology.kubernetes.io/zone
+        operator: In
+        values: ["us-west-2a"]
+      - key: topology.k8s.aws/zone-id
+        operator: In
+        values: ["usw2-az1"]
+  - price: ....
+    available: ....
+    requirements:
+      - key: karpenter.sh/capacity-type
+        operator: In
+        values: ["spot"]
+      - key: topology.kubernetes.io/zone
+        operator: In
+        values: ["us-west-2a"]
+      - key: topology.k8s.aws/zone-id
+        operator: In
+        values: ["usw2-az1"]
+```
+
+### Representing ODCR Available Instance Counts in Instance Type Offerings
+
+ODCRs (unlike spot and on-demand capacity) have much more defined, constrained capacity ceilings. For instance, in an extreme example, a user may select on a capacity reservation with only a single available instance but launch 10,000 pods that contain hostname anti-affinity. The scheduler would do work to determine that it needs to launch 10,000 instances for these pods; however, without any kind of cap on the number of times the capacity reservation offering could be used, the scheduler would think that it could launch 10,000 instances into the capacity reservation offering. 
+
+Attempting to launch this would result in a success for a single instance and an ICE error for the other 9,999. The next scheduling loop would remediate this, but this results in a lot of extra, unneeded work. 
+
+A better way to model this would be to track the available instance count as a numerical value associated with an instance type offering. In this modeling, the scheduler could count the number of simulated NodeClaims that might use the offering and know that it can't simulate NodeClaims into particular offerings once they hit their cap.
+
+Today, we already have an [`available` field](https://github.com/kubernetes-sigs/karpenter/blob/bcd33e924905588b1bdecd5413dc7b268370ec4c/pkg/cloudprovider/types.go#L236) attached to instance type offerings. This field is binary and only tells us whether the instance is or isn't available. With the introduction of ODCR offerings, we could extend this field to be an integer rather than a boolean. This would allow us to exactly represent the number of available instances that can be launched into the offering. Existing spot and on-demand offerings would model this `available` field as `MAX_INT` for current `true` values and `0` for `false` values.
```

Comment:
another alternative to MAX_INT is using an `*int` where nil means available, no limit. A bit weird though when you think about unavailable going from nil to 0.

Reply:
Yeah, I think MAX_INT is more natural here -- we're never going to launch MAX_INT instances